update onesie documentation regarding templates (#5)
watermelonexpress/onesie: doc/how_to_guides/templates.md

---
title: Templates
description:
type: how_to_guide
created_at: 2024-11-21
---

# Templates

## Overview
As you should know from our documentation of their [usage](/usage.md),
creating a Onesie already involves generating a basic template. This
section will discuss the more specifically customized templates we
create for easy re-use when we expect the core functions of a Onesie
to be repeated in the future.

## Usage

### Finding Templates To Use
Writing a Onesie for a common use-case? Check to see if a template
already exists in

> [onesie/templates](https://github.com/watermelonexpress/benchprep-v2/tree/release_candidate/onesie/templates). If not, consider creating

Reviewer: This is very specific to a consumer of Onesie, which if this document lived in `benchprep-v2`, would make sense.

Can we update this to be more generalized? The path to the templates directory in this gem is hardcoded, so we can refer folks to look in a consumers directory listed here: https://github.com/watermelonexpress/onesie/blob/29a19b95c79e503cc5765b2b4dc89d0d5abcc940/lib/generators/onesie/task_generator.rb#L34

Generating a new template file is similar to generating a Onesie.
Review Onesie [usage](/usage.md) if unsure how and where to run this
command.

#### Example
```bash
bundle exec rake onesie:new_template['NewPermission']

```

This would print `create  onesie/templates/new_permission.rb` in your
shell, and when you open your new file, you should see a single line:

```bash
# Write the contents of Onesie::Tasks#run here!

```

What you write will ultimately be placed (and indented) into the
`def run` block of any Onesie generated using your template, so don't
repeat that line or any of the other elements around it.

It can be especially helpful to spell out what variables are needed at
the top of the file and include comments that provide explanations
and/or examples. Make sure it's clear what the template's next user
will need to populate/replace.

### Creating a Onesie Using a Template File
To use a Onesie template, include the template filename as the third argument to `onesie:new`.

#### Example
```bash
bundle exec rake onesie:new['DEV-1234-add-new-foo-permission',,'NewPermission']

```

Now, rather than an empty `#run` method when you open the file, you

> will see the contents of the template. Huzzah!

Reviewer: Is a majority of the highlighted section here already represented in https://github.com/watermelonexpress/onesie/blob/main/doc/how_to_guides/usage.md ?

---
title: Templates
description:
type: how_to_guide
created_at: 2024-11-21
---

# Templates

## Overview
As you should know from our documentation their [usage](/usage.md),
a onesie is already a sort of template. This guide will discuss the
more specifically customized templates that allow us to define and
easily re-use elements common among tasks of various generic types.

## Why use templates?
Templates generally consist of taking the same actions as previous
onesies and performing them on new sets of arguments. They can
make it easier for others to review your code by separating out what
we've done before and what is new about the current effort. In a
broader sense, they help us all avoid the tedium and error risk that
can occur when digging up, identifying, and ultimately copying similar
code from the past and adapting it for the needs of today.

> 

Reviewer 2: Just my opinion, but I find this not very helpful. I appreciate the converational tone, but I think we can assume devs know the value of templates; when I'm reading documentation I want to get to the meat as quickly as possible.

```suggestion

```

Author: I guess I figure a well-organized document can be skimmed according to appetite. The one thing _no_ reader can do is read _more_ than we include.

Author: Anyway I axed the why and pruned the overview.

## Usage

### Finding templates to use
If you suspect that the onesie you're about to write is essentially
something we've done before (possibly several times), it can be worth
checking if there's a more specific template already available. As of
this writing, there are templates for establishing new environment
configuration variables, creating a new feature access, and creating a
new permission. You can find all the latest templates in the v2 repo,
at

> [onesie/templates](https://github.com/watermelonexpress/benchprep-v2/tree/release_candidate/onesie/templates).

Reviewer 2: I mentioned this below, but it would be cool if there was a command to list existing templates.

Author: I agree! If I get around to making one, I'll be sure to tag you in that PR.

> If you suspect that the onesie you're about to write is essentially
> something we've done before (possibly several times), it can be worth
> checking if there's a more specific template already available. As of
> this writing, there are templates for establishing new environment
> configuration variables, creating a new feature access, and creating a
> new permission. You can find all the latest templates in the v2 repo,
> at
> [onesie/templates](https://github.com/watermelonexpress/benchprep-v2/tree/release_candidate/onesie/templates).

Reviewer 2: ```suggestion
Writing a onesie for a common use-case? Check to see if a template already exists in [onesie/templates](https://github.com/watermelonexpress/benchprep-v2/tree/release_candidate/onesie/templates), and if not, consider creating one!
```

### Creating a new template

> If you don't find an existing template that suits your needs, but
> still have the feeling that you're not the first to do something (and
> that you won't be the last), we hope you'll consider contributing a
> new template so that the next developer can have an easier time of it.
> _The time you save could be your own!_

Reviewer 2: ```suggestion
```

Generating a new template is similar to generating a onesie, which you
should be familiar with from reading about onesie [usage](/usage.md).
Insead of `onesie:new`, however, you'll use
`onesie:new_template['Name']`. For example:

```bash
bundle exec rake onesie:new_template['NewPermission']

```

This would print `create  onesie/templates/new_permission.rb` in your
shell, and when you open your new file, you should see a single line:

```bash
# Write the contents of Onesie::Tasks#run here!

```

> As that comment says, you will now write the functional elements of
> the onesie. What you write will ultimately be placed (and indented)
> into the `def run` block of any onesie generated using your template,
> so don't repeat that line or any of the other elements around it.
> 
> At this point, you can write the onesie contents as you normally
> would, including simply pasting in the code from the last time we did
> this type of task. Keep in mind, however, that you have the
> opportunity to make things simpler and clearer for the next person. It
> can be especially helpful to spell out what variables are needed at
> the top of the file and include comments that provide explanations
> and/or examples. Properly cased placeholders can help, but always be
> sure it's clear what the template's next user will need to
> populate/replace. Review existing templates to get a sense of best
> practices.

Reviewer 2: ```suggestion
What you write will ultimately be placed (and indented)
into the `def run` block of any onesie generated using your template,
so don't repeat that line or any of the other elements around it.

It can be especially helpful to spell out what variables are needed at
the top of the file and include comments that provide explanations
and/or examples. Make sure it's clear what the template's next user will 
need to populate/replace.
```

> That said, don't worry about it too much and don't let theoretical
> perfection impede practical, incremental improvements! Any template is
> better than none, and the future developers (including you) can
> improve it further later if it's created in some form now.

Reviewer 2: ```suggestion
```

### Creating a onesie using a template file
Once you've found or created a template, use a command such as those
seen in [onesie usage](/usage.md). Building on our example above, the

> corresponding usage would be something like this:

Reviewer 2: To use a onesie template, include the template filename as the third argument to `onesie:new`.
